coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala

```diff
@@ -54,14 +57,66 @@ class QueryEngine(dataset: Dataset,
     }
   }
 
+  /**
+    * Converts Routes to ExecPlan
+    */
+  def routeExecPlanMapper(routes: Seq[Route], rootLogicalPlan: LogicalPlan, queryId: String, submitTime: Long,
+                          options: QueryOptions, spreadProvider: SpreadProvider): ExecPlan = {
+    val execPlans : Seq[ExecPlan]= routes.map { route =>
+      route match {
+        case route: LocalRoute => if (!route.asInstanceOf[LocalRoute].tr.isDefined)
+          genExecPlan(rootLogicalPlan, queryId, submitTime, options, spreadProvider)
+        else
+          genExecPlan(QueryRoutingPlanner.updateTimeLogicalPlan(rootLogicalPlan,
+            route.asInstanceOf[LocalRoute].tr.get),
+            queryId, submitTime, options, spreadProvider)
+
+        case route: RemoteRoute =>
+          val remoteLogicalPlan = route.asInstanceOf[RemoteRoute].tr.map(_ =>QueryRoutingPlanner.
+            updateTimeLogicalPlan(rootLogicalPlan,
+            route.asInstanceOf[RemoteRoute].tr.get)).getOrElse(rootLogicalPlan)
+
+          RemoteExec(queryId, route.dispatcher, dataset.ref, exec.RemoteExecParams(remoteLogicalPlan,
+            options.copy(processFailures = true)), submitTime)
+      }
+    }
+
+    if (execPlans.size == 1)
+      return execPlans(0)
+    else
+      // Stitch RemoteExec plan results with local using InProcessorDispatcher
+      StitchRvsExec(queryId, InProcessPlanDispatcher(dataset), execPlans)
+  }
+
   /**
     * Converts a LogicalPlan to the ExecPlan
     */
   def materialize(rootLogicalPlan: LogicalPlan,
-                  options: QueryOptions, spreadProvider: SpreadProvider = StaticSpreadProvider()): ExecPlan = {
+                  options: QueryOptions): ExecPlan = {
     val queryId = UUID.randomUUID().toString
+    val submitTime = System.currentTimeMillis()
+    val querySpreadProvider = options.spreadProvider.getOrElse(spreadProvider)
 
-    val materialized = walkLogicalPlanTree(rootLogicalPlan, queryId, System.currentTimeMillis(),
+    if (!QueryRoutingPlanner.isPeriodicSeriesPlan(rootLogicalPlan) || !options.processFailures)
+      return genExecPlan(rootLogicalPlan, queryId, submitTime, options, querySpreadProvider)
+
+      val failures: Seq[FailureTimeRange] = failureProvider.getFailures(dataset.ref,
+        QueryRoutingPlanner.getTimeFromLogicalPlan(logicalPlan = rootLogicalPlan)).
+        sortWith(_.timeRange.startInMillis < _.timeRange.startInMillis)
+
+    if(failures.isEmpty)
+      return genExecPlan(rootLogicalPlan, queryId, submitTime, options, querySpreadProvider)
```

Comment:
If return statement is avoidable by using if-else, please do. In this method, I do think it is.